main.tf

```diff
@@ -69,60 +69,126 @@ resource "aws_route_table_association" "this" {
+                "redshift:DescribeClusters",
+                "tag:GetResources"
+            ],
+            "Effect": "Allow",
+            "Resource": "*"
+        },
+        {
+            "Action": [
+                "lambda:AddPermission",
+                "lambda:CreateFunction",
+                "lambda:GetFunction",
+                "lambda:InvokeFunction",
+                "lambda:UpdateFunctionConfiguration"
+            ],
+            "Effect": "Allow",
+            "Resource": "arn:aws:lambda:*:*:function:SecretsManager*"
+        },
+        {
+            "Action": [
+                "serverlessrepo:CreateCloudFormationChangeSet",
+                "serverlessrepo:GetApplication"
+            ],
+            "Effect": "Allow",
+            "Resource": "arn:aws:serverlessrepo:*:*:applications/SecretsManager*"
+        },
+        {
+            "Action": [
+                "s3:GetObject"
+            ],
+            "Effect": "Allow",
+            "Resource": [
+                "arn:aws:s3:::awsserverlessrepo-changesets*",
+                "arn:aws:s3:::secrets-manager-rotation-apps-*/*"
+            ]
+        }
+    ]
+}
 
-  ingress {
-    description      = "HTTP from EC2 instance"
-    from_port        = 80
-    to_port          = 80
-    protocol         = "tcp"
-    cidr_blocks      = ["0.0.0.0/0"]
-  }
+  )
+}
+#create appropriate role for ec2 instance
+resource "aws_iam_role" "this" {
+  name = "ec2_instance_sm_role_terraform"
 
-  egress {
-    from_port        = 0
-    to_port          = 0
-    protocol         = "-1"
-    cidr_blocks      = ["0.0.0.0/0"]
-  }
+  # Terraform's "jsonencode" function converts a
+  # Terraform expression result to valid JSON syntax.
+  assume_role_policy = jsonencode({
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Action": [
+                "sts:AssumeRole"
+            ],
+            "Principal": {
+                "Service": [
+                    "ec2.amazonaws.com"
+                ]
+            }
+        }
+    ]
+})
+}
 
-  tags = {
-    Name = "terraform"
-  }
+#unite iam role and policy
+resource "aws_iam_policy_attachment" "this" {
+  name       = "ec2_attachment_SM"
+  roles      = [aws_iam_role.this.name]
+  policy_arn = aws_iam_policy.this.arn
 }
 
-#creating EC2 instance 
-resource "aws_instance" "web_server_test" {
-  ami                         = var.ami_type
-  instance_type               = var.instance_type
-  subnet_id                   = aws_subnet.this.id
-  vpc_security_group_ids      = [aws_security_group.this.id]
-  key_name                    = var.key_pair_ssh 
-  user_data                   = file("script-docker-install-rhel.sh")
-  associate_public_ip_address = true
+resource "aws_iam_instance_profile" "this" {
+  name = "ec2_sm_profile"
+  role = aws_iam_role.this.name
+}
 
-  tags = {
-    Name = "Terraform server"
-  }
+#creating EC2 instance 
+module "module_server_test" {
```

Comment:
SG groups works tightly with EC2 server. You can put these resources into one module. 
The idea of modules is to simplify the code.

Reply:
Yes, but here is an another task for creating SG on LB. In this case - I should create separate SG resource for LB?